File: app/src/main/java/ai/doc/netrunner/view/LiveCameraClassificationFragment.java
package ai.doc.netrunner.view;


import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProviders;
import androidx.databinding.DataBindingUtil;
import android.graphics.Bitmap;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import android.view.LayoutInflater;
import android.view.TextureView;
import android.view.View;
import android.view.ViewGroup;

import java.util.List;
import java.util.Map;

import ai.doc.netrunner.ModelRunner;
import ai.doc.netrunner.R;
import ai.doc.netrunner.databinding.FragmentLiveCameraClassificationBinding;
import ai.doc.tensorio.TIOUtilities.TIOClassificationHelper;

/**
 * A simple {@link Fragment} subclass.
 */

public class LiveCameraClassificationFragment extends LiveCameraFragment implements ModelRunner.ModelRunnerDataSource {
    private static final int RESULTS_TO_SHOW = 3;
    private static final int FILTER_STAGES = 3;
    private static final float FILTER_FACTOR = 0.4f;

    private MutableLiveData<String> latency = new MutableLiveData<>();
    private MutableLiveData<String> predictions = new MutableLiveData<>();
    private TextureView textureView;
    private float[][] filterLabelProbArray = null;

    public LiveCameraClassificationFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        FragmentLiveCameraClassificationBinding binding = DataBindingUtil.inflate(inflater, R.layout.fragment_live_camera_classification, container, false);
        binding.setFragment(this);
        binding.setLifecycleOwner(this);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        textureView = view.findViewById(R.id.texture);
    }

    @Override
    public Bitmap getNextInput(int size_x, int size_y) {
        return textureView.getBitmap(size_x, size_y);
    }

    @Override
    public void onResume() {
        super.onResume();
        startClassification();
    }

    @Override
    public void onPause() {
        super.closeCamera();
        stopClassification();
        super.onPause();
    }

    public void startClassification() {
        ClassificationViewModel vm = ViewModelProviders.of(getActivity()).get(ClassificationViewModel.class);

        vm.getModelRunner().startStreamClassification(this, (requestId, output, l) -> {
            Map<String, Float> classification = (Map<String, Float>)((Map<String,Object>)output).get("classification");
            List<Map.Entry<String, Float>> top5 = TIOClassificationHelper.topN(classification, RESULTS_TO_SHOW);
            String top5formatted = formattedResults(top5);

            // TODO: Apply smoothing filter

            predictions.postValue(top5formatted);
            latency.postValue(l + " ms");
        });
    }

    public void stopClassification() {
        ClassificationViewModel vm = ViewModelProviders.of(getActivity()).get(ClassificationViewModel.class);
        vm.getModelRunner().stopStreamClassification();
    }

    public LiveData<String> getLatency() {
        return latency;
    }

    public LiveData<String> getPredictions() {
        return predictions;
    }

    private String formattedResults(List<Map.Entry<String, Float>> results) {
        StringBuilder b = new StringBuilder();

        for (Map.Entry<String, Float> entry : results) {
            b.append(entry.getKey());
            b.append(": ");
            b.append(String.format("%.2f", entry.getValue()));
            b.append("\n");
        }

        b.setLength(b.length() - 1);

        return b.toString();
    }

    private void applyFilter(float[] result, int numLabels) {

        if (filterLabelProbArray == null || filterLabelProbArray[0].length != numLabels) {
            filterLabelProbArray = new float[FILTER_STAGES][numLabels];
        }

        // Low pass filter `labelProbArray` into the first stage of the filter.
        for (int j = 0; j < numLabels; ++j) {
            filterLabelProbArray[0][j] +=
                    FILTER_FACTOR * (result[j] - filterLabelProbArray[0][j]);
        }
        // Low pass filter each stage into the next.
        for (int i = 1; i < FILTER_STAGES; ++i) {
            for (int j = 0; j < numLabels; ++j) {
                filterLabelProbArray[i][j] +=
                        FILTER_FACTOR * (filterLabelProbArray[i - 1][j] - filterLabelProbArray[i][j]);
            }
        }

        // Copy the last stage filter output back to `labelProbArray`.
        System.arraycopy(filterLabelProbArray[FILTER_STAGES - 1], 0, result, 0, numLabels);
    }

}

File: app/src/main/java/ai/doc/netrunner/view/LiveCameraClassificationFragment.kt
package ai.doc.netrunner.view

import ai.doc.netrunner.ModelRunner.ModelRunnerDataSource
import ai.doc.netrunner.ModelRunner.ClassificationResultListener
import ai.doc.netrunner.R
import ai.doc.netrunner.databinding.FragmentLiveCameraClassificationBinding

import ai.doc.tensorio.TIOUtilities.TIOClassificationHelper

import android.graphics.Bitmap
import android.os.Bundle
import android.view.LayoutInflater
import android.view.TextureView
import android.view.View
import android.view.ViewGroup

import androidx.databinding.DataBindingUtil
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProviders

private const val RESULTS_TO_SHOW = 3
private const val FILTER_STAGES = 3
private const val FILTER_FACTOR = 0.4f

/**
 * A simple [Fragment] subclass.
 */

class LiveCameraClassificationFragment : LiveCameraFragment(), ModelRunnerDataSource, ClassificationResultListener {

    // UI

    private lateinit var textureView: TextureView

    // Live Data Variables

    private val _latency = MutableLiveData<String>()
    val latency: LiveData<String> = _latency

    private val _predictions = MutableLiveData<String>()
    val predictions: LiveData<String> = _predictions

    // View Model

    // requires fragment-ktx dependency
    // val viewModel: ClassificationViewModel by activityViewModels()

    private val viewModel: ClassificationViewModel by lazy {
        ViewModelProviders.of(requireActivity()).get(ClassificationViewModel::class.java)
    }

    private var filterLabelProbArray: Array<FloatArray>? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val binding: FragmentLiveCameraClassificationBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_live_camera_classification, container, false)
        binding.fragment = this
        binding.lifecycleOwner = this
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        textureView = view.findViewById(R.id.texture)
    }

    override fun getNextInput(): Bitmap? {
        return textureView.bitmap
    }

    override fun classificationResult(requestId: Int, output: Any?, l: Long) {
        val classification = (output as Map<String?, Any?>)["classification"] as Map<String, Float>?
        val top5 = TIOClassificationHelper.topN(classification, RESULTS_TO_SHOW)
        val top5formatted = formattedResults(top5)

        // TODO: Apply smoothing filter
        _predictions.postValue(top5formatted)
        _latency.postValue("$l ms")
    }

    // TODO: why are onResume and onPause here

    override fun onResume() {
        super.onResume()
        startClassification()
    }

    override fun onPause() {
        super.closeCamera()
        stopClassification()
        super.onPause()
    }

    fun startClassification() {
        viewModel.modelRunner.startStreamClassification(this) { requestId: Int, output: Any, l: Long ->
            val classification = (output as Map<String?, Any?>)["classification"] as Map<String, Float>?
            val top5 = TIOClassificationHelper.topN(classification, RESULTS_TO_SHOW)
            val top5formatted = formattedResults(top5)

            // TODO: Apply smoothing filter
            _predictions.postValue(top5formatted)
            _latency.postValue("$l ms")
        }
    }

    fun stopClassification() {
        viewModel.modelRunner.stopStreamClassification()
    }

    private fun formattedResults(results: List<Map.Entry<String, Float>>): String {
        val b = StringBuilder()
        for ((key, value) in results) {
            b.append(key)
            b.append(": ")
            b.append(String.format("%.2f", value))
            b.append("\n")
        }
        b.setLength(b.length - 1)
        return b.toString()
    }

    // TODO: Update exponential decay function

//    private fun applyFilter(result: FloatArray, numLabels: Int) {
//        if (filterLabelProbArray == null || filterLabelProbArray!![0].length != numLabels) {
//            filterLabelProbArray = Array(FILTER_STAGES) { FloatArray(numLabels) }
//        }
//
//        // Low pass filter `labelProbArray` into the first stage of the filter.
//        for (j in 0 until numLabels) {
//            filterLabelProbArray!![0][j] += FILTER_FACTOR * (result[j] - filterLabelProbArray!![0][j])
//        }
//        // Low pass filter each stage into the next.
//        for (i in 1 until FILTER_STAGES) {
//            for (j in 0 until numLabels) {
//                filterLabelProbArray!![i][j] += FILTER_FACTOR * (filterLabelProbArray!![i - 1][j] - filterLabelProbArray!![i][j])
//            }
//        }
//
//        // Copy the last stage filter output back to `labelProbArray`.
//        System.arraycopy(filterLabelProbArray!![FILTER_STAGES - 1], 0, result, 0, numLabels)
//    }
}